buildSrc/plugins/src/main/kotlin/org/jetbrains/kotlin/CoverageTest.kt

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.kotlin
+
+import groovy.lang.Closure
+import org.gradle.api.DefaultTask
+import org.gradle.api.Task
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.TaskAction
+
+/**
+ * Test task for -Xcoverage and -Xlibraries-to-cover flags. Requires a binary to be built by the Konan plugin
+ * with
+ * konanArtifacts {
+ *      program([binaryName], targets: [testTarget]) {
+ *          ...
+ *          extraOpts "-Xcoverage"/"-Xlibrary-to-cover=...", "-Xcoverage-file=$[profrawFile]"
+ *      }
+ *  }
+ * and a dependency set according to a pattern "run${binaryName}".
+ *
+ * @property numberOfCoveredFunctions Expected number of covered functions
+ * @property numberOfCoveredLines Expected number of covered lines in all functions
+ * @property binaryName Name of the produced binary
+ */
+open class CoverageTest : DefaultTask() {
+
+    private val target = project.testTarget()
+    private val platform = project.platformManager().platform(target)
+    private val llvmBin = "${platform.configurables.absoluteLlvmHome}/bin"
+
+    @Input
+    lateinit var binaryName: String
+
+    // TODO: Consider better metric.
+    @Input
+    var numberOfCoveredFunctions: Int? = null
+    @Input
+    var numberOfCoveredLines: Int? = null
+
+    val profrawFile: String by lazy {
+        "${project.buildDir.absolutePath}/$binaryName.profraw"
+    }
+
+    private val profdataFile: String by lazy {
+        "${project.buildDir.absolutePath}/$binaryName.profdata"
+    }
+
+    private val outputDir: String by lazy {
+        project.file(project.property("testOutputCoverage")!!).absolutePath
+    }
+
+    override fun configure(closure: Closure<Any>): Task {
+        super.configure(closure)
+        if (!(project.property("useCustomDist") as Boolean)) {
+            setRootDependency("${target}CrossDist", "${target}CrossDistRuntime", "commonDistRuntime", "distCompiler")
+        }
+        dependsOn(project.tasks.getByName("run$binaryName"))
+        return this
+    }
+
+    private fun setRootDependency(vararg s: String) = s.forEach { dependsOn(project.rootProject.tasks.getByName(it)) }
```

Comment:
Isn't this mean that all `crossdist`s will depend on the test task? The test task should depend on `crossDist` and build task